publish-mqtt/src/homie.rs

```diff
@@ -0,0 +1,204 @@
+use std::error::Error;
+use tokio::task::{self, JoinHandle};
+
+const HOMIE_VERSION: &str = "4.0";
+
+#[derive(Clone, Debug)]
+pub struct Property {
+    id: String,
+    name: String,
+    datatype: String,
+    unit: String,
+}
+
+impl Property {
+    pub fn new(id: &str, name: &str, datatype: &str, unit: &str) -> Property {
+        Property {
+            id: id.to_string(),
+            name: name.to_string(),
+            datatype: datatype.to_string(),
+            unit: unit.to_string(),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Node {
+    id: String,
+    name: String,
+    node_type: String,
+    properties: Vec<Property>,
+}
+
+impl Node {
+    pub fn new(id: String, name: String, node_type: String, properties: Vec<Property>) -> Node {
+        Node {
+            id,
+            name,
+            node_type,
+            properties,
+        }
+    }
+}
+
+pub(crate) struct HomieDevice {
+    requests_tx: Sender<Request>,
+    device_base: String,
+    device_name: String,
+    nodes: Vec<Node>,
+}
+
+impl HomieDevice {
+    pub async fn new(
+        device_base: &str,
+        device_name: &str,
+        mut mqtt_options: MqttOptions,
+    ) -> (
+        HomieDevice,
+        JoinHandle<Result<(), Box<dyn Error + Send + Sync>>>,
+    ) {
+        mqtt_options.set_last_will(LastWill {
+            topic: format!("{}/$state", device_base),
+            message: "lost".to_string(),
+            qos: QoS::AtLeastOnce,
+            retain: true,
+        });
+        let mut event_loop = EventLoop::new(mqtt_options, 10).await;
+
+        let homie = HomieDevice {
+            requests_tx: event_loop.handle(),
+            device_base: device_base.to_string(),
+            device_name: device_name.to_string(),
+            nodes: vec![],
+        };
+
+        let join_handle: JoinHandle<Result<(), Box<dyn Error + Send + Sync>>> =
+            task::spawn(async move {
+                loop {
+                    let (incoming, outgoing) = event_loop.poll().await?;
+                    log::trace!("Incoming = {:?}, Outgoing = {:?}", incoming, outgoing);
+                }
+            });
+
+        (homie, join_handle)
+    }
+
+    pub async fn start(&self) -> Result<(), SendError<Request>> {
+        publish_retained(
+            &self.requests_tx,
+            format!("{}/$homie", self.device_base),
+            HOMIE_VERSION,
+        )
+        .await?;
+        publish_retained(
+            &self.requests_tx,
+            format!("{}/$extensions", self.device_base),
+            "",
+        )
+        .await?;
+        publish_retained(
+            &self.requests_tx,
+            format!("{}/$name", self.device_base),
+            &self.device_name,
+        )
+        .await?;
+        publish_retained(
+            &self.requests_tx,
+            format!("{}/$state", self.device_base),
+            "init",
+        )
+        .await?;
+        Ok(())
```

Comment:
(observation) there is a school of thought that says:
```suggestion
        .await
```
if `publish_retained(...).await` returns `Result<(), _>`

I sort-of prefer it the way you've written it though.

Reply:
I did consider that, but I don't like it in this case because it spoils the consistency with the other statements in this method.